### PR TITLE
Fixes #54. All JS optimized. CDN added and fallback to local minified files.

### LINKED ISSUE #54: JS optimization
JQuery that is used is pretty old. And we are using local copy. It will optimize performance of web page quite a lot. 

If we can use Google cdn JQuery (fallback will be local one.)
And move scripts on bottom. That will not block page loading. 

Let me know what you think on this.

### PR DESCRIPTION
**Pull request by kunjee17**
All js minified and moved to bottom of the page. Vendor files are point to CDN and fallback to latest minified version of file. 

Currently minification is done manually if needed we can set up gulp or grunt and join it with FAKE.

### CODE REVIEW

**Comment by tpetricek** [comment]
This is looking great!

I think that doing minification automatically as part of the build is a great idea. Without that, it is likely that someone will send a PR changing the scripts and I won't know what to do with it :-).

I suppose that will need to invoke some tool, so we'll also need to make sure to get the tool from somewhere using Paket.

It would be also good if this worked in the automatic reloading - when you change the `js` file during development, the FAKE task should process it & serve a new version.

**Comment by kunjee17** [comment]
@tpetricek gulp/grunt is there and integration with FAKE is also there. I don't know what is the preference and how to add it. So, for now I have keep it as it is. If you want I can set up as I like?

And there is way it can be done using node packages just like that. Means whatever you wish.

**Comment by tpetricek** [comment]
I think it would be better to merge this PR when it comes with FAKE integration, otherwise we're in danger of getting out of sync.

**Comment by kunjee17** [comment]
@tpetricek give me a day. I ll get it done with FAKE. Hope you don't mind node module in your project?   I will mostly use [this](http://fsharp.github.io/FAKE/apidocs/fake-npmhelper.html) to do the trick.

**Comment by tpetricek** [comment]
As long as it builds on my & the Azure machine, it should be all good

**Comment by kunjee17** [comment]
:+1:

**Comment by kunjee17** [comment]
@tpetricek hey, I have made changes so every time people make changes to js in app folder in watch so one every changes it will minify file and put it in dist folder. That can be used in application. Please have a look and see if there is changes to be done as per convention you are following. Also check on your machine as it works on mine. :)

**Comment by Krzysztof-Cieslak** [comment]
1. Things like `jquery` and `bootstrap` can be installed as npm dependency, instead of putting them on GitHub.
2. Minification is OK, but having also bundling (i.e. making one JS and one CSS file) would be nice for deployed version. I suggest using [webpack](https://webpack.github.io/), but in JS world it's all matter of personal preferences ;)

**Comment by kunjee17** [comment]
@Krzysztof-Cieslak I wanted to go with minimum changes. And in current case files are referenced separately in html files. So, one is loading for one page and other is for another. I don't wanted to change code as of now so I keep them separate and just minify it. 

I agree with JQuery part. But again I am not sure I should go for bower / npm. Currently checked in files are also getting used as back up only. 

Basically all depends how much JavaScript magic we wanted to bring in. For few scripts I feel its just a overhead. It is all my personal two cents.

**Comment by kunjee17** [comment]
@tpetricek can we take this PR in? We can change things as people who wanted to work in front end come to conclusion to work on some specific way of doing things.

**Comment by tpetricek** [comment]
I just tried it and I get the following error on my machine:

```
Starting Target: JS
Node js web compilation thing
c:\Tomas\Public\web\FsSnip\fssnip-website\packages\Npm.js\tools\npm.cmd install
uglifyjs-folder@0.2.1 node_modules\uglifyjs-folder
├── extend@2.0.1
├── mkdirp@0.5.1 (minimist@0.0.8)
├── readdir@0.0.13 (q@1.0.1)
├── meow@1.0.0 (object-assign@1.0.0, minimist@1.2.0, camelcase-keys@1.0.0, indent-string@1.2.2)
└── uglify-js@2.6.1 (uglify-to-browserify@1.0.2, async@0.2.10, source-map@0.5.3, yargs@3.10.0)
c:\Tomas\Public\web\FsSnip\fssnip-website\packages\Npm.js\tools\npm.cmd run build

> fssnip@1.0.0 build c:\Tomas\Public\web\FsSnip\fssnip-website
> npm run build:js

'npm' is not recognized as an internal or external command,
operable program or batch file.

npm ERR! Windows_NT 6.3.9600
npm ERR! argv "c:\\Tomas\\Public\\web\\FsSnip\\fssnip-website\\packages\\Node.js\\node.exe" "c:\\Tom
as\\Public\\web\\FsSnip\\fssnip-website\\packages\\Npm.js\\tools\\node_modules\\npm\\bin\\npm-cli.js
" "run" "build"
npm ERR! node v5.3.0
npm ERR! npm  v2.13.1
npm ERR! code ELIFECYCLE
npm ERR! fssnip@1.0.0 build: `npm run build:js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the fssnip@1.0.0 build script 'npm run build:js'.
npm ERR! This is most likely a problem with the fssnip package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run build:js
npm ERR! You can get their info via:
npm ERR!     npm owner ls fssnip
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     c:\Tomas\Public\web\FsSnip\fssnip-website\npm-debug.log
Running build failed.
Error:
System.Exception: 'npm run build' task failed
   at Fake.NpmHelper.run(NpmParams npmParams) in C:\code\fake\src\app\FakeLib\NpmHelper.fs:line 59
   at FSI_0001.Build.clo@177-16.Invoke(Unit _arg5) in c:\Tomas\Public\web\FsSnip\fssnip-website\buil
d.fsx:line 182
   at Fake.TargetHelper.runSingleTarget(TargetTemplate`1 target) in C:\code\fake\src\app\FakeLib\Tar
getHelper.fs:line 483
```

**Comment by Krzysztof-Cieslak** [comment]
I guess, Scripts commands in `package.json` should be changed to use local (Paket) `node` and `npm` installation

**Comment by kunjee17** [comment]
@tpetricek @Krzysztof-Cieslak I guess that is not needed. Have you install paket. New paket is added for local node and npm?

**Comment by tpetricek** [comment]
I run `build.sh` in my Git shell...

**Comment by kunjee17** [comment]
@tpetricek does that restore paket also. I haven't checked with build.sh. But I needed to install using .paket\paket.exe install thing. Before I am able to run build.cmd.

**Comment by tpetricek** [comment]
Yes, it gets the latest version of `.paket/paket.exe` first. 

I do not have node installed, so it needs to use the ones from the `packages` folder.

**Comment by kunjee17** [comment]
@tpetricek can you try by installing paket manually first so we can drill down that if that is build.sh / cmd part I missed or node is crying. I also try on my end. But I have node install for environment. Let me check. 

**Update1**
I took a fresh copy of my repo and run build.cmd and things work.

**Comment by tpetricek** [comment]
@kunjee17 Do you have some existing `npm` in your path? I think the problem is that I don't:

```
> npm run build:js

'npm' is not recognized as an internal or external command,
operable program or batch file.
```

It is installed in `packages`, but the script calls the global one...

**Comment by kunjee17** [comment]
@tpetricek I do have node and npm in global path. :(. NpmHelper as per doc should use package one only. I don't know how to point to package one. Let me check if I can do anything.

Here is [sample](http://blog.2mas.xyz/npm-build-with-fsharp-fake/) I followed.

**Comment by Krzysztof-Cieslak** [comment]
`NpmHelper` is spawning right one (hence `c:\Tomas\Public\web\FsSnip\fssnip-website\packages\Npm.js\tools\npm.cmd run build`). The problem is with script defined in `package.json`.

**Comment by kunjee17** [comment]
@Krzysztof-Cieslak oh man. I guess there is path issue. I am giving local path for uglify folder. Sh**. Can you please help me update it. I m in Windows machine now. 

Its not path. I guess. Let me fix it.

**Comment by kunjee17** [comment]
@tpetricek I have changed the npm path. Can you please pull it latest and check it out. 

@Krzysztof-Cieslak thanks for pointer. It was indeed package.json path. 

Sorry for mess.

**Comment by tpetricek** [comment]
It is still failing, but I think it got further now!

```
The resulting target order is:
 - JS
Starting Target: JS
Node js web compilation thing
c:\Tomas\Public\web\FsSnip\fssnip-website\packages\Npm.js\tools\npm.cmd install
c:\Tomas\Public\web\FsSnip\fssnip-website\packages\Npm.js\tools\npm.cmd run build

> fssnip@1.0.0 build c:\Tomas\Public\web\FsSnip\fssnip-website
> packages\Npm.js\tools\npm.cmd run build:js


> fssnip@1.0.0 build:js c:\Tomas\Public\web\FsSnip\fssnip-website
> node node_modules/uglifyjs-folder/cli.js web/content/app/ -eo web/content/dist/

'node' is not recognized as an internal or external command,
operable program or batch file.

```

File: package.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "fssnip",
+  "version": "1.0.0",
+  "description": "F# Snippets - Web site\r ======================",
+  "main": "index.js",
+  "scripts": {
+    "build:js": "node node_modules/uglifyjs-folder/cli.js web/content/app/ -eo web/content/dist/",
```
Comment (Krzysztof-Cieslak) [inline package.json]: `packages\\Node.js\\node.exe` instead of `node`

**Comment by kunjee17** [comment]
@tpetricek @Krzysztof-Cieslak node path is also updated. Have a look now. It should work.

**Comment by tpetricek** [comment]
All works fine now, thanks :-)

I did one more tweak, which is to add the `web/content/dist` folder to `.gitignore`, because I think we do not need the generated files in the repo - they would just make the diffs more noisy.

**Comment by kunjee17** [comment]
@tpetricek thank. Finally it happened. Sorry for lots of to and fro. 

BTW how things looks on server? Is it look better than before?
